BDArmory/UI/BDArmoryAIGUI.cs

```diff
@@ -296,42 +265,51 @@ void GetAIEditor()
-                    {
-                        List<BDGenericAIBase> aiQuery = aiType switch
-                        {
-                            ActiveAIType.PilotAI => p.FindModulesImplementing<BDModulePilotAI>().ConvertAll(ai => ai as BDGenericAIBase),
-                            ActiveAIType.SurfaceAI => p.FindModulesImplementing<BDModuleSurfaceAI>().ConvertAll(ai => ai as BDGenericAIBase),
-                            ActiveAIType.VTOLAI => p.FindModulesImplementing<BDModuleVTOLAI>().ConvertAll(ai => ai as BDGenericAIBase),
-                            ActiveAIType.OrbitalAI => p.FindModulesImplementing<BDModuleOrbitalAI>().ConvertAll(ai => ai as BDGenericAIBase),
-                            _ => null
-                        };
-                        if (aiQuery == null || aiQuery.Count == 0) continue; // None of this type found.
-                        foreach (var ai in aiQuery)
-                        {
-                            if (ai == null) continue;
-                            if (ai == ActiveAI) yield break; // We found the current active AI!
-                            activeAIType = aiType;
-                            ActiveAI = ai;
-                            SetInputFields(aiType);
-                            SetChooseOptionSliders();
-                            yield break;
-                        }
-                    }
+                    var rootPart = ship.Parts.First(); while (rootPart.parent != null) rootPart = rootPart.parent;
+                    AIs.Sort((m1, m2) => ProximityToRoot(m1.part, rootPart).CompareTo(ProximityToRoot(m2.part, rootPart))); // Same sorting as ActiveController with no active AIs.
+                    if (ActiveAI == AIs.First() as BDGenericAIBase) yield break; // It's the same AI, do nothing.
+                    ActiveAI = AIs.First() as BDGenericAIBase; // Switch back to the primary AI.
+                    activeAIType = ActiveAI.aiType;
+                    SetInputFields(activeAIType);
+                    SetChooseOptionSliders();
+                    yield break;
                 }
             }
 
             // No AIs were found, clear everything.
-            activeAIType = ActiveAIType.None;
+            activeAIType = AIType.None;
             ActiveAI = null;
             inputFields = null;
+            AIs.Clear();
+        }
+
+        /// <summary>
+        /// Get the proximity to the root part.
+        /// </summary>
+        /// <param name="part"></param>
+        /// <returns>Proximity to the root part or int.MaxValue if not connected to the root part.</returns>
+        public static int ProximityToRoot(Part part, Part root)
+        {
+            int proximity = 0;
+            Part currentPart = part;
+            while (currentPart != null && currentPart != root)
+            {
+                currentPart = currentPart.parent;
+                ++proximity;
+            }
+            if (currentPart == null)
```

Comment:
Would it be sensible to have a provision for stopping proximity checks if AI is behind a docking port or decoupler (similar to how MMGs determine what parts are the MMG and not parent craft)?

Reply:
That's a possible option, but it means that if the AI / WM on the mothership gets destroyed, then the AI on the parasite fighters can't take over.
Also, something silly like a chain of "cars" ([Test - multicraft.zip](https://github.com/user-attachments/files/21330909/Test.-.multicraft.zip)) with different types of AIs on wouldn't be able to switch between them without undocking first, which would limit players options a little.